source/snapshot/iterator.go

```diff
@@ -100,6 +100,54 @@ func (i *Iterator) Next(ctx context.Context) (opencdc.Record, error) {
 	}
 }
 
+// NextN takes and returns up to n records from the queue. NextN is allowed to
+// block until either at least one record is available or the context gets canceled.
+func (i *Iterator) NextN(ctx context.Context, n int) ([]opencdc.Record, error) {
+	if n <= 0 {
+		return nil, fmt.Errorf("n must be greater than 0, got %d", n)
+	}
+
+	var records []opencdc.Record
```

Comment:
There's an optimization we can do here by pre-allocating the slice, otherwise we're creating more allocations than we should, especially for big batches.

Now, what size should we pick? An easy solution would be to just allocate `n` elements and expect that we'll fill the slice up. Of course that will waste space if we don't actually fill the slice up. So I'm thinking that we could store the last batch size in the iterator and use that to allocate the slice, if it's smaller than `n`. The assumption here is that we will mostly see batches of the same size. For instance, if records are steadily streaming and batches are filled, we'll constantly create slices of the full batch. Then, if the stream dies down and we don't have a lot to process, the batch size will decrease and we'll allocate smaller slices.

```suggestion
	records := make([]opencdc.Record, min(i.lastBatchSize, n))
	...
	// before returning the records store the batch size
	i.lastBatchSize = len(records)
	return records, nil
```

Or do you see a better strategy?

Reply:
>  Of course that will waste space if we don't actually fill the slice up. 

How much space would actually be wasted? 

There's another thing that plays a role and that is the `snapshot.fetchSize` parameter (which I think probably can be removed, because it we have `sdk.batch.size`). While the iterator is active, we can always expect the database to return that amount of records (except for the last fetch),

Reply:
@hariso good point. I have actually been confused by the two whenever I used this connector as a source. This change would require deprecating the field, making use only of the SDK parameter, but I'll see how to approach this in a way that doesn't increase much scope.

Reply:
This is a bit more nuanced than the CDC iterator, because records are coming in the fetch size defined in the config. As it stands, each record from that cursor fetch is fed 1 by 1 into the channel. If there is caching at that level, you essentially need to take N and not have to pre-allocate.

Reply:
> There's another thing that plays a role and that is the `snapshot.fetchSize` parameter (which I think probably can be removed, because it we have `sdk.batch.size`). While the iterator is active, we can always expect the database to return that amount of records (except for the last fetch),

@hariso I'm unclear about this, the fetchSize is the cursor fetch size, e.g. how many rows to get from Postgres, this is primarily used to optimize retrieval of rows. Removing that will result in `select * from table` which may or may not perform better, but something we should try, but somewhat tangential to this PR.

Reply:
> How much space would actually be wasted?

If you have batches of 10k records, but fill them up with 1 record, you're wasting 99.99% of the space you're allocating 🙃 Using `unsafe.Sizeof(opencdc.Record{})` the empty record is determined to be 104 bytes, so allocating 9999 of them is just shy of 1MB. In the worst case, when records trickle in just slow enough, this can happen tens or hundreds of times per second, putting quite a lot of pressure on the garbage collector.

Reply:
@lyuboxa what I meant is to keep the `FETCH`, but not have a `fetchSize` parameter in the connector. Instead, we should use `FETCH sdk.batch.size`.

Reply:
@hariso I think this will be a bit confusing, since `sdk.batch.size` is a general config, while fetch size is specific for the snapshot.

Reply:
Why does it need to be specific? When I find `sdk.batch.size: 100` in a source connector's configuration, I expect the connector to read 100 records at a time, which is the fetch size in snapshots, or no?

Reply:
Because it does not apply to CDC. Its specific to postgres and it may become irrelevant if we decided to move to `select * from table` ¯\_(ツ)_/¯

Reply:
@hariso and I had a chat, the TL;DR is that `sdk.batch.size` is what conduit passes to `ReadN` and that may change on configuration, thus allocating a slice with `n` capacity based on the `ReadN` argument makes sense.

```go
records := make([]opencdc.Record, 0, sdkBatchSize)
```

The compiler has no idea how to allocate this on the stack, so it will go on the heap anyway. 

The append to that should be fairly efficient and you can either return the full slice or something like `records[:nread]` but that doesn't matter, the slice will have the same capacity. You need to get that from the source batch configuration from the config.

What @lovromazgon is suggesting will be needed if we are returning `[]opencdc.Record` but I don't think we need it in this implementation.

Reply:
Based on my understanding, the first size `n` should always be based on the batch size configured by the SDK anyways (`ReadN`) so the allocation would still be `n`. i.e.: I agree with the last comment by @lyuboxa . Unless I'm missing something else here.

Reply:
Allocating size `n` is fine with me, better than not allocating! 👍